### PR TITLE
Address assign and request tweaks in response to editor discussions

### LINKED ISSUE #54: Timing interaction between Address-Request and Address-Assign capsule and the Connect-IP request
So there appear to exist some downside with the timing of the interactions when one want to use Address-Request capsule. So the basic problem is simply that the proxy will assign an address prior to the Address-Request capsule has been received, thus preventing the proxy from processing the request in the right context.  

So lets assume that the client has already created one connect-IP request with a limited scope and been assigned IP-A. Then it want to make an additional request for another limited scope proxying and would like to get the same address assigned. Due to the separation between the HTTP request for the Connect-IP and the capsule Address-Request a server can actually process the whole HTTP request prior to receiving the Address-Request. Resulting potentially in that another IP address assignment (IP-B) be provided to the client, despite it asking for IP-A to be assigned also to the request. 

So does the HTTP request needs an indicator parameter to tell the proxy to wait for one or more "request" capsules from the client before processing the request?

### LINKED ISSUE #66: Should there be a way to say "please assign me an address but I don't care which" ?
Implementing this on the client I was wondering if there should be a way for the client to ask for the server for an address without specifying what address it wants. This could allow a VPN client to mention that it wants an IPv6 address or not.

If we think this could be useful, we could encode it as `ADDRESS_ASSIGN(IPv4-0.0.0.0/32)` or `ADDRESS_ASSIGN(IPv6-::/128)`.

### PR DESCRIPTION
**Pull request by DavidSchinazi**
Fixes #54 
Fixes #66

### CODE REVIEW

File: draft-ietf-masque-connect-ip.md [hunk 1]
```diff
@@ -280,35 +289,47 @@ address, the receiver of this capsule is allowed to send packets from any source
...
 
 In some deployments of CONNECT-IP, an endpoint needs to be assigned an address
 by its peer before it knows what source address to set on its own packets. For
 example, in the Remote Access case ({{example-remote}}) the client cannot send
-IP packets until it knows what address to use. In these deployments, endpoints
-need to send ADDRESS_ASSIGN capsules to allow their peers to send traffic.
+IP packets until it knows what address to use. In these deployments, the
+endpoint that is expecting an address assignment MUST send an ADDRESS_ASSIGN
+capsule. This isn't required if the endpoint does not need any address
+assignment, for example when it is configured out of band with static addresses.
```
Comment (achernya) [inline draft-ietf-masque-connect-ip.md:302]: nit: out-of-band

File: draft-ietf-masque-connect-ip.md [hunk 2]
```diff
@@ -280,35 +289,47 @@ address, the receiver of this capsule is allowed to send packets from any source
...
+
+While ADDRESS_ASSIGN capsules are commonly sent in response to ADDRESS_REQUEST
+capsules, endpoints MAY send ADDRESS_ASSIGN capsules unprompted.
```
Comment (achernya) [inline draft-ietf-masque-connect-ip.md:305]: Do we actually want this text? I agree that this is permitted per our discussion as "send an ADDRESS_REQUEST if you need an ADDRESS_ASSIGN" does not preclude unsolicited ADDRESS_ASSIGN, but I can't think of a use case here that requires this clarification.

**Review by achernya — APPROVED** [review]
Thanks! I left 2 minor editorial questions, but feel free to merge as-is.

File: draft-ietf-masque-connect-ip.md (hunk 1 above)
Comment (tfpauly) [reply draft-ietf-masque-connect-ip.md:302]:
```suggestion
assignment, for example when it is configured out-of-band with static addresses.
```

File: draft-ietf-masque-connect-ip.md (hunk 2 above)
Comment (tfpauly) [reply draft-ietf-masque-connect-ip.md:305]: I think it's nice to have a MAY here just to make sure people don't read it as a MUST NOT.

File: draft-ietf-masque-connect-ip.md [hunk 3]
```diff
@@ -280,35 +289,47 @@ address, the receiver of this capsule is allowed to send packets from any source
...
 ~~~
 ADDRESS_REQUEST Capsule {
   Type (i) = ADDRESS_REQUEST,
   Length (i),
+  Requested Address (..) ...,
+}
+~~~
+{: #addr-req-format title="ADDRESS_REQUEST Capsule Format"}
+
+The ADDRESS_ASSIGN capsule contains a sequence of Requested Addresses.
```
Comment (tfpauly) [inline draft-ietf-masque-connect-ip.md:323]:
```suggestion
The ADDRESS_REQUEST capsule contains a sequence of Requested Addresses.
```

**Review by tfpauly — APPROVED** [review]

**Review by gloinul — APPROVED** [review]

File: draft-ietf-masque-connect-ip.md [hunk 4]
```diff
@@ -280,35 +289,47 @@ address, the receiver of this capsule is allowed to send packets from any source
...
+all of the assigned addresses or prefixes can be used. For example, multiple
+Assigned Addresses are necessary to assign both IPv4 and IPv6 addresses.
 
 In some deployments of CONNECT-IP, an endpoint needs to be assigned an address
 by its peer before it knows what source address to set on its own packets. For
 example, in the Remote Access case ({{example-remote}}) the client cannot send
-IP packets until it knows what address to use. In these deployments, endpoints
-need to send ADDRESS_ASSIGN capsules to allow their peers to send traffic.
+IP packets until it knows what address to use. In these deployments, the
+endpoint that is expecting an address assignment MUST send an ADDRESS_ASSIGN
```
[the remark this answers is not in the document]
Comment (DavidSchinazi) [reply draft-ietf-masque-connect-ip.md:300]: Oh good catch, thanks!
```suggestion
endpoint that is expecting an address assignment MUST send an ADDRESS_REQUEST
```

File: draft-ietf-masque-connect-ip.md [hunk 5]
```diff
@@ -833,9 +874,9 @@ Capsule Types" registry created by {{HTTP-DGRAM}}:
 
 |  Value   |        Type         |     Description     |   Reference   |
 |:---------|---------------------|:--------------------|:--------------|
-| 0xfff100 |   ADDRESS_ASSIGN    | Address Assignment  | This Document |
-| 0xfff101 |   ADDRESS_REQUEST   | Address Request     | This Document |
-| 0xfff102 | ROUTE_ADVERTISEMENT | Route Advertisement | This Document |
+| 0xfff200 |   ADDRESS_ASSIGN    | Address Assignment  | This Document |
+| 0xfff201 |   ADDRESS_REQUEST   | Address Request     | This Document |
```
Comment (martinthomson) [inline draft-ietf-masque-connect-ip.md:878]: you had an opportunity to pick something using rand(), but you didn't take it?

Comment (DavidSchinazi) [reply draft-ietf-masque-connect-ip.md]: Fixed

**Review by martinthomson — COMMENTED** [review]
I'll reiterate my suggestion from the meeting: it would be better to have IP-version-specific capsule types for each capsule.  It saves a byte and also avoids problems with version mismatches.

**Comment by DavidSchinazi** [comment]
@martinthomson let's discuss that on #72. I'd like to merge this PR first then decide on a plan for #72 before submitting the next revision to the datatracker

File: draft-ietf-masque-connect-ip.md [hunk 6]
```diff
@@ -243,21 +243,31 @@ these new capsules.
...
 ~~~
 ADDRESS_ASSIGN Capsule {
   Type (i) = ADDRESS_ASSIGN,
   Length (i),
+  Assigned Address (..) ...,
+}
+~~~
+{: #addr-assign-format title="ADDRESS_ASSIGN Capsule Format"}
+
+The ADDRESS_ASSIGN capsule contains a sequence of Assigned Addresses.
```
Comment (LPardue) [inline draft-ietf-masque-connect-ip.md:261]: sequence of 0 or more?

Comment (DavidSchinazi) [reply draft-ietf-masque-connect-ip.md:261]:
```suggestion
The ADDRESS_ASSIGN capsule contains a sequence of zero or more
Assigned Addresses.
```

**Pull request merged**